docs/architecture/adr-029-fee-grant-module.md

```diff
@@ -0,0 +1,142 @@
+barrier to adoption.
+
+For instance, when proper permissions are setup, someone may temporarily delegate the ability to vote on proposals to
+a "burner" account that is stored on a mobile phone with only minimal security.
+
+Other use cases include workers tracking items in a supply chain or farmers submitting field data for analytics
+or compliance purposes.
+
+For all of these use cases, UX would be significantly enhanced by obviating the need for these accounts to always
+maintain the appropriate fee balance. This is especially true if we wanted to achieve enterprise adoption for something
+like supply chain tracking.
+
+While one solution would be to have a service that fills up these accounts automatically with the appropriate fees, a better UX
+would be provided by allowing these accounts to pull from a common fee pool account with proper spending limits.
+A single pool would reduce the churn of making lots of small "fill up" transactions and also more effectively leverages
+the resources of the organization setting up the pool.
+
+## Decision
+
+As a solution we propose a module, `x/feegrant` which allows one account, the "granter" to grant another account, the "grantee"
+an allowance to spend the granter's account balance for fees within certain well-defined limits.
+
+Fee allowances are defined by the extensible `FeeAllowanceI` interface:
+
+```go
+type FeeAllowanceI {
+    // Accept can use fee payment requested as well as timestamp/height of the current block
+ 	// to determine whether or not to process this. This is checked in
+ 	// Keeper.UseGrantedFees and the return values should match how it is handled there.
+ 	//
+ 	// If it returns an error, the fee payment is rejected, otherwise it is accepted.
+ 	// The FeeAllowance implementation is expected to update it's internal state
+ 	// and will be saved again after an acceptance.
+ 	//
+ 	// If remove is true (regardless of the error), the FeeAllowance will be deleted from storage
+ 	// (eg. when it is used up). (See call to RevokeFeeAllowance in Keeper.UseGrantedFees)
+ 	Accept(fee sdk.Coins, blockTime time.Time, blockHeight int64) (remove bool, err error)
+}
+```
+
+Two basic fee allowance types, `BasicFeeAllowance` and `PeriodicFeeAllowance` are defined to support known use cases:
+
+```proto
+// BasicFeeAllowance implements FeeAllowance with a one-time grant of tokens
+// that optionally expires. The delegatee can use up to SpendLimit to cover fees.
+message BasicFeeAllowance {
+     repeated cosmos_sdk.v1.Coin spend_limit = 1;
+     ExpiresAt expiration = 2;
+}
+
+// PeriodicFeeAllowance extends FeeAllowance to allow for both a maximum cap,
+// as well as a limit per time period.
+message PeriodicFeeAllowance {
+     BasicFeeAllowance basic = 1;
+     Duration period = 2;
+     repeated cosmos_sdk.v1.Coin period_spend_limit = 3;
+     repeated cosmos_sdk.v1.Coin period_can_spend = 4;
```

Comment:
maybe add some doc about these fields like it was the case in https://gist.github.com/aaronc/b60628017352df5983791cad30babe56#core-feeallowance-types?

Reply:
I added those comments and in the process also updated the `ExpiresAt` and `Duration` structs a bit because they seemed a bit sloppy. Now these use `oneof` which seems more logical. Can you take a look again @blushi ?

Reply:
> Now these use oneof which seems more logical. Can you take a look again @blushi @amaurymartiny ?

yeah makes sense